src/cursor/abstract_cursor.ts

```diff
@@ -282,7 +283,17 @@ export abstract class AbstractCursor<
 
   /** Returns current buffered documents */
   readBufferedDocuments(number?: number): TSchema[] {
-    return this[kDocuments].splice(0, number ?? this[kDocuments].length);
+    const bufferedDocs: TSchema[] = [];
+    const documentsToRead = number ?? this[kDocuments].length;
+
+    for (let count = 0; count < documentsToRead; count++) {
+      const document = this[kDocuments].shift();
```

Comment:
> we `splice` ed the array, which did not modify

That's not correct, [splice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) modifies the array you use it on, so we were removing elements from the `kDocuments` array and we still are now.

> Why are we filtering out null values now? is that to handle the sentinel?

No the sentinel isn't exposed outside the list, it's because of typescript, even though `documentsToRead` is bound to the size of the list and you know, and I know, and everyone who can read the code knows it cannot be null... typescript cannot know :sadcat: `shift` returns `T | null` so I must either cast away the problem or conditionally check its existence.

Reply:
We did not have adequate tests for this API, I've added more and fixed the bounding